### PR TITLE
Add SwapperCalleePsm

### PR DESCRIPTION
**Pull request by telome**
Some open questions:

- do we want to automatically `fill()` the PSM buffer when there is not enough dai available to be bought? Currently leaning towards no as large swaps by facilitators can call `fill()` manually and small swaps via `StableSwapper` may assume that keepers are regularly filling the PSM buffer. Swaps via `StableSwapper` may also be less common than for the UniV3 callee given that there is no slippage risk to mitigate here.

- do we want to use the `data` of the `swapCallback()` to pass the PSM address (and therefore support multiple PSMs more easily)? Currently leaning towards no as only the USDC PSM is currently considered and passing the PSM address as `data` implies additional swap costs.

- do we need a token recovery method in case the wrong gem is sent to the callee? Probably not necessary as only swappers are going to be warded callers of `swapCallback()`.

### CODE REVIEW

File: test/mocks/PsmMock.sol [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+pragma solidity ^0.8.16;
+
+interface GemLike {
+    function approve(address spender, uint256 value) external;
+    function transfer(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+    function decimals() external view returns (uint8);
+}
+
+contract KegMock {
```
Comment (sunbreak1211) [inline test/mocks/PsmMock.sol]:
```suggestion
contract PocketMock {
```

File: test/mocks/PsmMock.sol [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    modifier auth() {
+        require(wards[msg.sender] == 1, "PsmMock/not-authorized");
+        _;
+    }
+
+    constructor(address dai_, address gem_) {
+        dai = dai_;
+        gem = gem_;
+        to18ConversionFactor = 10**(18 - GemLike(gem_).decimals());
+        keg = address(new KegMock(gem_));
```
Comment (sunbreak1211) [inline test/mocks/PsmMock.sol]:
```suggestion
        pocket = address(new PocketMock(gem_));
```

File: test/mocks/PsmMock.sol (hunk 1 above)
Comment (sunbreak1211) [reply test/mocks/PsmMock.sol]: I'd actually argue that for the Mock you do not even need the `Pocket`.

File: test/mocks/PsmMock.sol [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+}
+
+contract PsmMock {
+    mapping(address => uint256) public wards;
+
+    address public immutable dai;
+    address public immutable gem;
+    uint256 public immutable to18ConversionFactor;
+    address public immutable keg;
```
Comment (sunbreak1211) [inline test/mocks/PsmMock.sol]:
```suggestion
    address public immutable pocket;
```

File: test/mocks/PsmMock.sol [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function sellGemNoFee(address usr, uint256 gemAmt) external auth returns (uint256 daiOutWad) {
+        daiOutWad = gemAmt * to18ConversionFactor;
+
+        GemLike(gem).transferFrom(msg.sender, keg, gemAmt);
```
Comment (sunbreak1211) [inline test/mocks/PsmMock.sol]:
```suggestion
        GemLike(gem).transferFrom(msg.sender, pocket, gemAmt);
```

File: test/mocks/PsmMock.sol [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        GemLike(dai).transfer(usr, daiOutWad);
+
+        emit SellGem(usr, gemAmt, 0);
+    }
+
+    function buyGemNoFee(address usr, uint256 gemAmt) external auth returns (uint256 daiInWad) {
+        daiInWad = gemAmt * to18ConversionFactor;
+
+        GemLike(dai).transferFrom(msg.sender, address(this), daiInWad);
+        GemLike(gem).transferFrom(keg, usr, gemAmt);
```
Comment (sunbreak1211) [inline test/mocks/PsmMock.sol]:
```suggestion
        GemLike(gem).transferFrom(pocket, usr, gemAmt);
```

File: src/funnels/callees/SwapperCalleePsm.sol [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+
+    constructor(address _psm) {
+        psm = _psm;
+        gem = PsmLike(psm).gem();
+        GemLike(PsmLike(psm).dai()).approve(address(psm), type(uint256).max);
+        GemLike(gem).approve(address(psm), type(uint256).max);
+        to18ConversionFactor = 10 ** (18 - GemLike(gem).decimals());
```
Comment (sunbreak1211) [inline src/funnels/callees/SwapperCalleePsm.sol:46]: FYI this is now exposed in the PSM. I'm fine keeping this way if preferred though.

File: test/mocks/PsmMock.sol (hunk 1 above)
Comment (telome) [reply test/mocks/PsmMock.sol]: Agreed, I removed it from the mock.

File: src/funnels/callees/SwapperCalleePsm.sol (hunk 6 above)
Comment (telome) [reply src/funnels/callees/SwapperCalleePsm.sol:46]: Right. Leaning towards keeping it as-is as it's a bit more explicit that way.

File: src/funnels/callees/SwapperCalleePsm.sol [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+pragma solidity ^0.8.16;
+
+interface GemLike {
+    function approve(address, uint256) external;
+    function decimals() external view returns (uint8);
+}
+
+interface PsmLike {
+    function sellGemNoFee(address usr, uint256 gemAmt) external returns (uint256 daiOutWad);
+    function buyGemNoFee(address usr, uint256 gemAmt) external returns (uint256 daiInWad);
```
Comment (sunbreak1211) [inline src/funnels/callees/SwapperCalleePsm.sol]:
```suggestion
    function sellGemNoFee(address, uint256) external returns (uint256);
    function buyGemNoFee(address, uint256) external returns (uint256);
```

File: test/mocks/PsmMock.sol [hunk 8]
```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.16;
+
+interface GemLike {
+    function approve(address spender, uint256 value) external;
```
Comment (sunbreak1211) [inline test/mocks/PsmMock.sol]:
```suggestion
    function approve(address, uint256) external;
```

File: src/funnels/callees/SwapperCalleePsm.sol (hunk 6 above)
Comment (sunbreak1211) [reply src/funnels/callees/SwapperCalleePsm.sol:46]: Sounds good!

**Review by sunbreak1211 — APPROVED** [review]
**do we want to automatically fill() the PSM buffer when there is not enough dai available to be bought?...**
I'd keep it simple as it is.

**do we want to use the data of the swapCallback() to pass the PSM address (and therefore support multiple PSMs more easily)?**
I like as it is.

**do we need a token recovery method in case the wrong gem is sent to the callee? Probably not necessary as only swappers are going to be warded callers of swapCallback().**
Yeah, if following the regular path, this should not happen, as otherwise it would revert. We can't ofc prevent that anyone is sending tokens, but I think we shouldn't prevent that.

LGTM.

**Review by oldchili — APPROVED** [review]
LGTM

**Pull request merged**